### PR TITLE
Feature/input group proposal

### PR DESCRIPTION
**Pull request by kioan000**
# Pyxis Component 6.0.0
## AIM

This proposal tries to:
- solve the grouping of more form input fields and render them properly (new FormFieldGroup type)
- evolve the previous API in a composable reusable way
- expose the whole validation process with advanced validation functions field by field, allowing the user to control when validate a field

## WARNING
- this proposal needs pyxis 2.1.0 (pull request: https://github.com/primait/pyxis-npm/pull/25)
- this proposal is **BREAKING**, you'll need to change your code accordingly to the new API

### Validation type
Specific validations have been dropped in favour of a generic validation constructor that exposes 
- severity level (Warning or Error)
- validation message
- validation function: `(FormModel -> Bool)` in positive logic (True means correct, False means wrong).
- validation Warnings will be painted ONLY when no Error Validations are triggered

### FormFieldList
Now when you need to render multiple fields in a row you should configure a `FormFieldList `.
FormFieldList can be created through `fieldListConfig` that accepts:
- a label
- a list of configured FormFields
- a list of Validations

FormFieldList adds a logic layer on Validation handling: 
- All the validations message (both FormField and FormFieldList) will be painted by FormFieldList in the same area
- When a List Error/Warning is present then all the fields borders turn red/yellow
- When a Field Error/Warning is present only triggered field border will turn red/yellow

### InputGroup
Now a input text field with an appended or prepended icon element is called InputGroup, you can create an InputGroup using append/prepend function that accepts a List(Html -> msg) and a FormField

### FormState and ValidationVisibilityPolicy
Form now have 3 native states, `Pristine | Touched | Submitted`
Now when you initialize a form you must specify which policy the form should adopt to display validations: 
- `WhenSubmitted`: form will paint validations only after you have explicitly set his state as submitted
- `Always`: form render always display validation messages. Use this configuration if you need a heavily custom show policy that depends on your own form state's (basically when Always the message visibility depends only by provided validation function)

### Tooltip rendering
Now you can add a Tooltip to a FormField or a FieldList and it will be rendered properly aligned in your Form using `addTooltipToFieldWhen` and `addTooltipToFieldListWhen`

### pureHtmlConfig
- now pureHtmlConfig produces an higly customizable FormField that allows user to specific field aspect and is wrapped in a generic a-form-field wrapper and prints out validation messages
```elm
staticHtmlField: FormField FormData Msg
staticHtmlField =
    Form.pureHtmlConfig
         -- slug
        "always-error-field"
        -- render content
        [ p [] [ text "You shall not pass" ]
        ]
        -- validations
        [ FormValidation.config FormValidation.Error
            (\formData -> False)
            "Fly, You Fools!"
        ]
```

### New fluid composable API
- Now you can add fields on the form with a fluid setter style like in the example (note that there is a specific adder form each of FormField, FormFieldList
```elm
{-| Somewhere in your View.elm
 your form rendering function now can be more readable and fluid
-}
div [ class "a-container directionColumn" ]
[ formConfig
            |> Form.addField Config.username
            |> Form.addFieldList Config.formFieldList
            |> Form.addField (Config.dateOfBirth data datePickerIcon)
            |> Form.addCustomRow (div [] [ text "Some fancy text in the middle of your form" ])
            |> Form.render data
]
...
{-| In your FormField configurators, now each field is properly typed and readable
-}
username : FormField FormData Msg
username =
    Form.textConfig
        --Slug
        "user_name"
        --Label
        (Just "Username")
        -- Data attributes
        [ minlength 3, maxlength 12 ]
        --FormData accessor
        .username
        --FormEvent mappings
        [ Event.onInput (UpdateField Username) ]
        --Form Validations
        [ FormValidation.config FormValidation.Error
            (\formData -> Maybe.withDefault False <| Maybe.map ((<) 3 << String.length) formData.username)
            "Username must be greater than 3 digits"
        ]

formFieldList : FormFieldList FormData Msg
formFieldList =
    Form.fieldListConfig
        -- Field List label
        "Username & Password"
        -- Field List FormFields
        [ username, password False ]
        -- Field List own validations
        [ FormValidation.config FormValidation.Warning
            (\formData -> not (formData.username == formData.password))
            "Username and password shouldn't be equal"
        ]

dateOfBirth : FormData -> Html Msg -> Form.InputGroup FormData Msg
dateOfBirth { isVisibleDP, dateOfBirthDP } appendable =
    Form.datepickerConfig
        "date_of_birth"
        (Just "Date of Birth")
        []
        .dateOfBirth
        (UpdateDatePicker DateOfBirth)
        [ Event.onInput (UpdateField DateOfBirth) ]
        dateOfBirthDP
        isVisibleDP
        [ FormValidation.config FormValidation.Error
            (\formData -> not (formData.dateOfBirth == Nothing))
            "You must select a date"
        ]
        |> Form.append [ appendable ]
        |> addTooltipToFieldWhen tooltipIsVisible dateOfBirthTooltip
...
dateOfBirthTooltip =
    Tooltip.upConfig 
           [ class "date-of-birth-tooltip" ]
           [ text "Tooltip on a Field"]
```

- To build an InputGroup you simply have to append or prepend some Html Msg to it with append/prepend function (it's still a valid form field)
- To add something in your Form that is not supposed to be a field (but a generic form row) you can now use `addCustomRow`function that adds your own render function directly in the body of the form without a wrapping
- To add a Tooltip on your field you can use addTooltipToFieldWhen (or addTooltipToFieldListWhen in case of FormFieldList)

### CODE REVIEW

File: src/Prima/Pyxis/Helpers.elm [hunk 1]
```diff
@@ -48,3 +50,17 @@ renderIf condition html =
...
+renderListIf : Bool -> List (Html msg) -> List (Html msg)
+renderListIf condition html =
+    if condition then
+        html
+
+    else
+        [ text "" ]
+
+
+flip : (a -> b -> c) -> b -> a -> c
```
Comment (matteosister) [inline src/Prima/Pyxis/Helpers.elm]: :cupid:

File: src/Prima/Pyxis/Form.elm [hunk 2]
```diff
@@ -143,161 +159,44 @@ type FormState
...
+    , tooltip : Maybe (Tooltip.Config msg)
+    }
 
 
 {-| Represents the configuration of a single form field.
 -}
 type FormField model msg
-    = FormField (FormFieldConfig model msg)
+    = SingleField (FormFieldConfig model msg)
+    | InputGroupField (InputGroup model msg)
```
Comment (alessiosantangelo) [inline src/Prima/Pyxis/Form.elm]: Qui rinominare: 
`SingleField (FormFieldConfig model msg)` --> `Field (FieldConfig model msg)`
`InputGroupField (InputGroup model msg)` --> `InputGroupField (InputGroupFieldConfig model msg)`

File: src/Prima/Pyxis/Form.elm [hunk 3]
```diff
@@ -1256,30 +1548,31 @@ renderRadioOption model ({ reader, slug, label, options, attrs } as config) inde
...
 
 
-renderCheckbox : model -> CheckboxConfig model msg -> List (Validation model) -> List (Html msg)
-renderCheckbox model ({ slug, label, options } as config) validations =
-    (List.concat << List.indexedMap (\index option -> renderCheckboxOption model config index option)) options
+renderCheckbox : CheckboxConfig model msg -> List (Html msg)
+renderCheckbox ({ slug, label, options } as config) =
+    [ div
+        [ class "a-form-field__checkbox-options" ]
+        ((List.concat << List.indexedMap (\index option -> renderCheckboxOption config index option)) options)
```
Comment (alessiosantangelo) [inline src/Prima/Pyxis/Form.elm]: Credo che questo si possa riscrivere: 
`((List.concat << List.indexedMap (renderCheckboxOption config)) options)`
(eliminando la closure)

File: src/Prima/Pyxis/Form.elm [hunk 4]
```diff
@@ -1601,384 +1814,366 @@ renderPureHtml { content } =
...
-        ( Expression (SeverityLevel Error) _ _, FormFieldSelectConfig _ _ ) ->
-            True
 
-        ( Expression (SeverityLevel Error) _ _, FormFieldTextareaConfig _ _ ) ->
-            True
+{-| Checks if a given field have been provided with some input
+-}
+fieldIsTouched : model -> FormField model msg -> Bool
+fieldIsTouched model formField =
+    not <| fieldIsPristine model <| formField
```
Comment (alessiosantangelo) [inline src/Prima/Pyxis/Form.elm]:
```
fieldIsTouched : model -> FormField model msg -> Bool
fieldIsTouched model =
   (not << fieldIsPristine model)
```

**Pull request merged**